### PR TITLE
Add :consolidate_protocols option to Mix.install/2

### LINKED ISSUE #10778: Mix.install should consolidate protocols
In most cases, we can assume new implementations are not going to be defined and this will give a boost to performance. We can support a consolidate_protocols: false flag for those who want to disable it. The consolidate_protocols option likely does not have to be in the cache key.

/cc @wojtekmach

### PR DESCRIPTION
**Pull request by wojtekmach**
closes #10778

### CODE REVIEW

**Comment by wojtekmach** [comment]
I'm getting warnings when running tests so any advice how to remove them would be appreciated!

```
% elixir lib/mix/test/mix_test.exs
Excluding tags: [windows: true]

..warning: redefining module InstallTest (current version loaded from /private/var/folders/qp/9xwx1_p50v9bhgj_8d_t62hr0000gn/T/mix_installs/elixir-1.12.0-dev-erts-11.1.8/b4a72b8aa7c4e645090a079ea889e308/_build/dev/lib/install_test/ebin/Elixir.InstallTest.beam)
  lib/install_test.ex:1

warning: redefining module InstallTest.Protocol (current version loaded from /private/var/folders/qp/9xwx1_p50v9bhgj_8d_t62hr0000gn/T/mix_installs/elixir-1.12.0-dev-erts-11.1.8/b4a72b8aa7c4e645090a079ea889e308/_build/dev/lib/mix_install/consolidated/Elixir.InstallTest.Protocol.beam)
  lib/install_test.ex:7

...warning: redefining module InstallTest (current version loaded from /private/var/folders/qp/9xwx1_p50v9bhgj_8d_t62hr0000gn/T/mix_installs/elixir-1.12.0-dev-erts-11.1.8/e73684e9f78939eca9a7d5584b968f56/_build/dev/lib/install_test/ebin/Elixir.InstallTest.beam)
  lib/install_test.ex:1

warning: redefining module InstallTest.Protocol (current version loaded from /private/var/folders/qp/9xwx1_p50v9bhgj_8d_t62hr0000gn/T/mix_installs/elixir-1.12.0-dev-erts-11.1.8/e73684e9f78939eca9a7d5584b968f56/_build/dev/lib/mix_install/consolidated/Elixir.InstallTest.Protocol.beam)
  lib/install_test.ex:7
```

File: lib/mix/test/mix_test.exs [hunk 1]
```diff
@@ -20,51 +20,98 @@ defmodule MixTest do
...
-        ]
+      assert Protocol.consolidated?(InstallTest.Protocol)
+
+      assert_received {:mix_shell, :info, ["==> install_test"]}
+      assert_received {:mix_shell, :info, ["Compiling 1 file (.ex)"]}
+      assert_received {:mix_shell, :info, ["Generated install_test app"]}
+      refute_received _
+
+      started_apps = Enum.map(Application.started_applications(), &elem(&1, 0))
+      assert :install_test in started_apps
```
Comment (josevalim) [inline lib/mix/test/mix_test.exs:40]:
```suggestion
      assert List.keyfind(Application.started_applications(), :install_test, 0)
```

**Comment by wojtekmach** [comment]
Updated!

File: lib/mix/lib/mix.ex [hunk 2]
```diff
@@ -492,6 +492,9 @@ defmodule Mix do
     * `:verbose` - if `true`, prints additional debugging information
       (Default: `false`)
 
+    * `:consolidate_protocols` - if `true`, runs protocol
+      consolidation via the `mix compile.protocols` task. (Default: `true`)
```
Comment (josevalim) [inline lib/mix/lib/mix.ex:496]:
```suggestion
      consolidation via the `mix compile.protocols` task (Default: `true`)
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**